src/webgpu/api/validation/encoding/encoder_state.spec.ts

```diff
@@ -114,3 +114,45 @@ g.test('call_after_successful_finish')
       encoder.finish();
     }
   });
+
+g.test('pass_end_none')
+  .desc(
+    `
+  Test that ending a {compute,render} pass without ending the passes generates a validation error.
+  `
+  )
+  .paramsSubcasesOnly(u => u.combine('passType', ['compute', 'render']).combine('endCount', [0, 1]))
+  .fn(async t => {
+    const { passType, endCount } = t.params;
+
+    const encoder = t.device.createCommandEncoder();
+
+    const pass = passType === 'compute' ? encoder.beginComputePass() : t.beginRenderPass(encoder);
+
+    for (let i = 0; i < endCount; ++i) {
+      pass.end();
+    }
+
+    t.expectValidationError(() => {
+      encoder.finish();
+    }, endCount === 0);
+  });
+
+g.test('pass_end_twice')
+  .desc('Test that ending a {compute,render} pass twice generates a validation error.')
+  .paramsSubcasesOnly(u => u.combine('passType', ['compute', 'render']).combine('endCount', [1, 2]))
+  .fn(async t => {
+    const { passType, endCount } = t.params;
+
+    const encoder = t.device.createCommandEncoder();
+
+    const pass = passType === 'compute' ? encoder.beginComputePass() : t.beginRenderPass(encoder);
+
+    for (let i = 0; i < endCount; ++i) {
+      pass.end();
+    }
+
+    t.expectValidationError(() => {
+      encoder.finish();
+    }, endCount > 1);
```

Comment:
There should be a validation error in the second pass.end() call, and then there shouldn't be one in finish.

```ts
pass.end();
if (endTwice) {
  t.expectValidationError(() => {
    pass.end();
  });
}

encoder.finish();
```

Chromium doesn't implement this yet, so the test will fail right now.

Reply:
Ah, sorry for misunderstanding this. I apply this to the test and a fail happens as we expect.